solidity/test/rewards/TestECDSARewardsDistributorEscrow.js

```diff
@@ -0,0 +1,137 @@
+      await token.approveAndCall(escrow.address, allTokens, "0x0", {
+        from: owner,
+      })
+    })
+
+    it("can not be called by non-owner", async () => {
+      await expectRevert(
+        escrow.allocateInterval(merkleRoot, amount, {from: thirdParty}),
+        "Ownable: caller is not the owner"
+      )
+    })
+
+    it("can be called by owner", async () => {
+      await escrow.allocateInterval(merkleRoot, amount, {from: owner})
+      // ok, no reverts
+    })
+
+    it("allocates reward distribution", async () => {
+      await escrow.allocateInterval(merkleRoot, amount, {from: owner})
+
+      const eventList = await rewardsDistributor.getPastEvents(
+        "RewardsAllocated",
+        {
+          fromBlock: 0,
+          toBlock: "latest",
+        }
+      )
+
+      expect(eventList.length).to.equal(1, "incorrect number of emitted events")
+      const event = eventList[0].returnValues
+      expect(event.merkleRoot).to.equal(merkleRoot, "unexpected merkle root")
+      expect(event.amount).to.eq.BN(amount)
+    })
+
+    it("allocates multiple reward distributions", async () => {
+      const merkleRoot2 =
+        "0xa7418520411d369b511eabb10ffb214c72b521ca0f6bd021fa83d9c47e65227e"
+      const amount2 = web3.utils.toBN(1337)
+
+      await escrow.allocateInterval(merkleRoot, amount, {from: owner})
+      await escrow.allocateInterval(merkleRoot2, amount2, {from: owner})
+
+      const eventList = await rewardsDistributor.getPastEvents(
+        "RewardsAllocated",
+        {
+          fromBlock: 0,
+          toBlock: "latest",
+        }
+      )
+
+      expect(eventList.length).to.equal(2, "incorrect number of emitted events")
+      const event1 = eventList[0].returnValues
+      expect(event1.merkleRoot).to.equal(merkleRoot, "unexpected merkle root")
+      expect(event1.amount).to.eq.BN(amount)
+      const event2 = eventList[1].returnValues
+      expect(event2.merkleRoot).to.equal(merkleRoot2, "unexpected merkle root")
+      expect(event2.amount).to.eq.BN(amount2)
+
+      expect(await token.balanceOf(rewardsDistributor.address)).to.eq.BN(
+        amount.add(amount2)
+      )
+    })
+  })
+
+  describe("funding", async () => {
+    it("can be done from phased escrow", async () => {
+      const fundingEscrow = await PhasedEscrow.new(token.address, {from: owner})
+      const allTokens = await token.balanceOf(owner)
+      await token.approveAndCall(fundingEscrow.address, allTokens, "0x0", {
+        from: owner,
+      })
+
+      const beneficiary = await ECDSARewardsEscrowBeneficiary.new(
+        token.address,
+        escrow.address,
+        {from: owner}
+      )
+      await beneficiary.transferOwnership(fundingEscrow.address, {from: owner})
+      await fundingEscrow.setBeneficiary(beneficiary.address, {from: owner})
+
+      await fundingEscrow.withdraw(allTokens, {from: owner})
+      expect(await token.balanceOf(escrow.address)).to.eq.BN(allTokens)
+    })
```

Comment:
What do you think about continue with a transfer in this test to the point when the funds are in `ECDSARewardsDistributorEscrow` contract? This way we connect all the dots end 2 end.

Reply:
I am not sure about this one. If we do that, we'll repeat the test from the previous `describe`.

This test ends with `expect(await token.balanceOf(escrow.address)).to.eq.BN(allTokens)` and it is a starting point of `describe("allocateInterval"`.

So maybe what we can do here is:
- switch the order of those two describes
- add check in `allocateInterval`'s `beforeEach` making sure all tokens were transferred (checking the same as this assertion)?

Reply:
Yes, sounds good. My point is just to show the flow of funds from `PhasedEscrow` to `ECDSARewardsDistributorEscrow` in e2e test so when reading you can clearly see what goes where step by step. Switching the order and adding a check will help with understanding the flow.

Reply:
Makes sense 👍 Please check f973c08ee19f4040b2f97a254786faa59129e130 to see if we achieved this goal.

Reply:
lgtm!